pkg/instrumentation/auto/annotation.go

```diff
@@ -0,0 +1,245 @@
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"golang.org/x/exp/maps"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aws/amazon-cloudwatch-agent-operator/pkg/instrumentation"
+)
+
+const (
+	autoAnnotatePrefix     = "cloudwatch.aws.amazon.com/auto-annotate-"
+	defaultAnnotationValue = "true"
+)
+
+// AnnotationMutators has an AnnotationMutator resource name
+type AnnotationMutators struct {
+	client              client.Client
+	logger              logr.Logger
+	namespaceMutators   map[string]instrumentation.AnnotationMutator
+	deploymentMutators  map[string]instrumentation.AnnotationMutator
+	daemonSetMutators   map[string]instrumentation.AnnotationMutator
+	statefulSetMutators map[string]instrumentation.AnnotationMutator
+	defaultMutator      instrumentation.AnnotationMutator
+}
+
+// MutateAll runs the mutators for each of the configured resources.
+func (m *AnnotationMutators) MutateAll(ctx context.Context) {
+	m.MutateNamespaces(ctx)
+	m.MutateDeployments(ctx)
+	m.MutateDaemonSets(ctx)
+	m.MutateStatefulSets(ctx)
+}
+
+// MutateNamespaces lists all namespaces and runs MutateNamespace on each.
+func (m *AnnotationMutators) MutateNamespaces(ctx context.Context) {
+	namespaces := &corev1.NamespaceList{}
+	if err := m.client.List(ctx, namespaces); err != nil {
+		m.logger.Error(err, "Unable to list namespaces")
+		return
+	}
+
+	for _, namespace := range namespaces.Items {
+		m.MutateNamespace(ctx, namespace)
+	}
+}
+
+// MutateNamespace modifies a single namespace's annotations using the configured mutators.
+func (m *AnnotationMutators) MutateNamespace(ctx context.Context, namespace corev1.Namespace) {
+	mutator, ok := m.namespaceMutators[namespace.Name]
+	if !ok {
+		mutator = m.defaultMutator
+	}
+	if !mutator.Mutate(namespace.GetObjectMeta()) {
+		return
+	}
+	if err := m.client.Update(ctx, &namespace); err != nil {
+		m.logger.Error(err, "Unable to send update", "kind", namespace.Kind, "name", namespace.Name)
+	}
+}
+
+// MutateDeployments lists all deployments and runs MutateDeployment on each.
+func (m *AnnotationMutators) MutateDeployments(ctx context.Context) {
+	deployments := &appsv1.DeploymentList{}
+	if err := m.client.List(ctx, deployments); err != nil {
+		m.logger.Error(err, "Unable to list deployments")
+		return
+	}
+	for _, deployment := range deployments.Items {
+		m.MutateDeployment(ctx, deployment)
+	}
+}
+
+// MutateDeployment modifies a single deployment's pod template spec annotations using the configured mutators.
+func (m *AnnotationMutators) MutateDeployment(ctx context.Context, deployment appsv1.Deployment) {
+	name := namespacedName(deployment.GetObjectMeta())
+	mutator, ok := m.deploymentMutators[name]
+	if !ok {
+		mutator = m.defaultMutator
+	}
+	if !mutator.Mutate(deployment.Spec.Template.GetObjectMeta()) {
+		return
+	}
+	if err := m.client.Update(ctx, &deployment); err != nil {
```

Comment:
If we're expecting to reuse this for the webhooks, the updates likely should be done outside of this function and the function should likely return the deployment and whether it was mutated.

Reply:
Though it looks like the webhooks don't care if the mutators made any changes, so maybe it doesn't need to return if it mutated the resource.